### PR TITLE
Generative module

### PR DESCRIPTION
**Pull request by dawidsroczyk**

### CODE REVIEW

**Review by domser2002 — COMMENTED** [review]
Will physical graph be used outside of generative proof machine? If not, better to keep it in one directory as a independent module

File: gtd_generative_proof_machine/inc/generative_restriction.h [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+{
+    int *path;
+    int *pathBool;
+    int numElems;
+    PathNode *next;
+} PathNode;
+
+typedef struct RestrictionResult
+{
+    int modified;
```
Comment (domser2002) [inline gtd_generative_proof_machine/inc/generative_restriction.h:22]: for boolean use bool from stdbool or uint8_t (usually equivalent in C), no need to use 32 bits here

File: gtd_generative_proof_machine/inc/generative_restriction.h [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+    int contradictionFound;
```
Comment (domser2002) [inline gtd_generative_proof_machine/inc/generative_restriction.h:23]: for boolean use bool from stdbool or uint8_t (usually equivalent in C), no need to use 32 bits here

File: gtd_generative_proof_machine/inc/generative_restriction.h [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+} RestrictionParameters;
+
+RestrictionParameters *initialize_restriction_parameters(void);
+void destroy_restriction_parameters(RestrictionParameters *params);
+PathNode *find_k_paths(Graph *graph, int k);
+GenerativeRestriction *create_restriction_object(RestrictionResult *(*condition)(Graph *graph, RestrictionParameters *params), RestrictionParameters *params);
+int delete_restriction_object(GenerativeRestriction *restriction);
+RestrictionResult *validate_restriction(Graph *graph, GenerativeRestriction *restriction);
+
+GenerativeRestriction *create_max_degree_restriction(int maxVertices, GenerativeProofMachine *machine);
```
Comment (domser2002) [inline gtd_generative_proof_machine/inc/generative_restriction.h:40]: consider introducing restriction type (best option is to move to common and reuse FactType enum, not every type needs to be supported both in generative machine and fact tree), then use one public constructor for GenerativeRestriction class with type argument

File: gtd_generative_proof_machine/inc/proof_tree.h [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+{
+    const char *message;
+    ProofTree *subtree;
+    ProofNode *next;
+} ProofNode;
+
+ProofNode *create_proof_node(void);
+ProofTree *create_proof_tree(void);
+void append_proof_node(ProofTree *tree, ProofNode *node);
+void write_proof_node(ProofNode *node, FILE *outFile, int depth);
```
Comment (domser2002) [inline gtd_generative_proof_machine/inc/proof_tree.h:23]: Will this function be called by anyone else than write_proof_tree? If not, remove the declaration from header and make it static

File: gtd_generative_proof_machine/inc/test_generative_proof_machine.h [hunk 5]
```diff
@@ -0,0 +1,18 @@
+#ifndef TEST_GENERATIVE_PROOF_MACHINE_H
+#define TEST_GENERATIVE_PROOF_MACHINE_H
+
+#include "physical_graph.h"
+#include "generative_restriction.h"
+#include "generative_proof_machine.h"
+#include <assert.h>
+
+void test_generative_proof_machine(void);
```
Comment (domser2002) [inline gtd_generative_proof_machine/inc/test_generative_proof_machine.h:9]: Again I am not sure if all those declarations are needed in header

File: gtd_generative_proof_machine/src/generative_proof_machine.c [hunk 6]
```diff
@@ -0,0 +1,152 @@
...
+
+/**
+ * \brief constructor for GenerativeProofMachine class
+ * \param numRestrictions number of generative restrictions
+ * \param startGraph start graph
+ * \returns pointer to a newly created GenerativeProofMachine
+ */
+GenerativeProofMachine *create_proof_machine(int numRestrictions, Graph *startGraph)
+{
+    GenerativeProofMachine *machine = gtd_malloc(sizeof(GenerativeProofMachine));
```
Comment (domser2002) [inline gtd_generative_proof_machine/src/generative_proof_machine.c:23]: good practice is to always cast type returned from malloc (does not have to be changed everywhere now, do not waste time)

File: gtd_generative_proof_machine/inc/generative_restriction.h [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+#include "physical_graph.h"
+#include "generative_proof_machine.h"
+
+typedef struct GenerativeRestriction GenerativeRestriction;
+typedef struct intNode intNode;
+typedef struct PathNode PathNode;
+
+typedef struct PathNode
+{
+    int *path;
```
Comment (domser2002) [inline gtd_generative_proof_machine/inc/generative_restriction.h:14]: good practice is to always specify int size and sign f.e. int32_t, uint8_t from stdint.h (does not have to be changed everywhere now, do not waste time)

File: gtd_generative_proof_machine/src/generative_proof_machine.c [hunk 8]
```diff
@@ -0,0 +1,152 @@
...
+    int i = 0;
+    while (i < machine->numRestrictions)
+    {
+        GenerativeRestriction *restriction = machine->restrictions[i];
+        Graph *graph = machine->graph;
+        RestrictionResult *result = validate_restriction(graph, restriction);
+
+        if (result->contradictionFound)
+        {
+            // TODO: Implement proper memory cleanup
```
Comment (domser2002) [inline gtd_generative_proof_machine/src/generative_proof_machine.c:62]: If it does not take a lot of time, let's not merge with obvious memory leaks

File: gtd_generative_proof_machine/src/generative_proof_machine.c [hunk 9]
```diff
@@ -0,0 +1,152 @@
...
+
+    return 1;
+}
+
+/**
+ * \brief function for executing generative proof machine
+ * \param machine pointer to the GenerativeProofMachine to execute
+ * \returns 1 if contradictionw was found, 0 otherwise
+ */
+int execute_generative_proof_machine(GenerativeProofMachine *machine)
```
Comment (domser2002) [inline gtd_generative_proof_machine/src/generative_proof_machine.c:51]: bool or uint8_t

File: gtd_physical_graph/src/physical_graph.c [hunk 10]
```diff
@@ -0,0 +1,198 @@
...
+
+    for(int r = 0; r < rows; ++r)
+    {
+        matrix[r] = gtd_malloc(cols * sizeof(char));
+    }
+
+    return matrix;
+}
+
+int int_pow(int base, int exp)
```
Comment (domser2002) [inline gtd_physical_graph/src/physical_graph.c:190]: Should it be in this file? Consider moving it to common or if it is static (used only in one file) make it static and remove declaration from header. There is also pow function in math.h (which we are already linking anyway)

File: gtd_test/unit_test.c [hunk 11]
```diff
@@ -1,10 +1,15 @@
...
 #include <stdio.h>
 #include "test_fact_tree.h"
+#include "test_physical_graph.h"
+#include "test_generative_proof_machine.h"
+#define TEST_MODE
 
 int main(void) {
     test_mode = true;
     test_fact_tree();
+    test_physical_graph();
```
Comment (domser2002) [inline gtd_test/unit_test.c:11]: only those 2 functions should be in headers (as mentioned previously)

File: gtd_generative_proof_machine/inc/proof_tree.h (hunk 4 above)
Comment (dawidsroczyk) [reply gtd_generative_proof_machine/inc/proof_tree.h:23]: yes, all of these functions are used outside of proof_tree.c (eg. in generative_restriction.c and generative_proof_machine.c)

File: gtd_generative_proof_machine/inc/generative_restriction.h (hunk 3 above)
Comment (dawidsroczyk) [reply gtd_generative_proof_machine/inc/generative_restriction.h:40]: it is not a simple fix. I will put it on todo list for the next pr

**Review by domser2002 — APPROVED** [review]

**Pull request merged**